### PR TITLE
improve partial opaque inference

### LINKED ISSUE #45780: Inference quality regression with opaque closures
In 1.7:
```
julia> struct Foo
           oc::Core.OpaqueClosure{Tuple{}}
       end

julia> f() = Val{Foo(@Base.Experimental.opaque ()->1).oc()}()
f (generic function with 1 method)

julia> @code_typed f()
CodeInfo(
1 ─     return $(QuoteNode(Val{1}()))
) => Val{1}
```
On master:
```
julia> @code_typed f()
CodeInfo(
1 ─ %1 = $(Expr(:new_opaque_closure, Tuple{}, Union{}, Int64, opaque closure(...) in Main))::Core.OpaqueClosure{Tuple{}}
│   %2 = (%1)()::Any
│   %3 = Core.apply_type(Main.Val, %2)::Type{Val{_A}} where _A
│   %4 = %new(%3)::Val
└──      return %4
) => Val
```

### PR DESCRIPTION
**Pull request by oscardssmith**
closes https://github.com/JuliaLang/julia/issues/45780

### CODE REVIEW

File: test/compiler/inference.jl [hunk 1]
```diff
@@ -4250,3 +4250,9 @@ let # NOTE make sure this toplevel chunk doesn't contain any local binding
     while xcond end
 end
 @test !xcond
+
+struct Issue45780
+    oc::Core.OpaqueClosure{Tuple{}}
+end
+f45780() = Val{Issue45780(@Base.Experimental.opaque ()->1).oc()}()
+@inferred @code_typed f45780()
```
Comment (Keno) [inline test/compiler/inference.jl:4258]:
```suggestion
@test (@inferred @code_typed f45780()) == Val{1}()
```

**Comment by Keno** [comment]
Could use a more extensive commit message, at least referencing the commit that added the `tmeet` call in the abstract interpretation for `:new`.

File: test/compiler/inference.jl [hunk 2]
```diff
@@ -4250,3 +4250,9 @@ let # NOTE make sure this toplevel chunk doesn't contain any local binding
     while xcond end
 end
 @test !xcond
+
+struct Issue45780
+    oc::Core.OpaqueClosure{Tuple{}}
+end
+f45780() = Val{Issue45780(@Base.Experimental.opaque ()->1).oc()}()
+@test (@inferred @code_typed f45780()) == Val{1}()
```
Comment (aviatesk) [inline test/compiler/inference.jl:4258]:
```suggestion
@test only(Base.return_types(f45780)) == Val{1}()
```

Comment (aviatesk) [reply test/compiler/inference.jl:4258]: That `@inferred` essentially checks if the return type-stability of `code_typed`, but not that of `f45780`.
Also note that `@code_typed` returns an IR too.

Comment (oscardssmith) [reply test/compiler/inference.jl:4258]: oops, I meant to check `@test (@inferred f45780()) == Val{1}()`. Would that be the same?

Comment (aviatesk) [reply test/compiler/inference.jl:4258]: Yeah, your option is better as it also includes actual execution also.

**Review by Keno — APPROVED** [review]

**Review by vtjnash — APPROVED** [review]

**Pull request merged**